python-package/xgboost/spark/core.py

```diff
@@ -886,6 +891,121 @@ def _get_xgb_parameters(
+                    "Stage-level scheduling in xgboost requires spark.executor.cores, "
+                    "spark.executor.resource.gpu.amount to be set."
+                )
+                return True
+
+            if int(executor_cores) == 1:
+                # there will be only 1 task running at any time.
+                self.logger.warning(
+                    "Stage-level scheduling in xgboost requires spark.executor.cores > 1 "
+                )
+                return True
+
+            if int(executor_gpus) > 1:
+                # For spark.executor.resource.gpu.amount > 1, we suppose user knows how to configure
+                # to make xgboost run successfully.
+                #
+                self.logger.warning(
+                    "Stage-level scheduling in xgboost will not work "
+                    "when spark.executor.resource.gpu.amount>1"
+                )
+                return True
+
+            task_gpu_amount = sc.getConf().get("spark.task.resource.gpu.amount")
+
+            if task_gpu_amount is None:
+                # The ETL tasks will not grab a gpu when spark.task.resource.gpu.amount is not set,
+                # but with stage-level scheduling, we can make training task grab the gpu.
+                return False
+
+            if float(task_gpu_amount) == float(executor_gpus):
+                self.logger.warning(
+                    "The configuration of cores (exec=%s, task=%s, runnable tasks=1) will "
+                    "result in wasted resources due to resource gpu limiting the number of "
+                    "runnable tasks per executor to: 1. Please adjust your configuration.",
+                    executor_gpus,
+                    task_gpu_amount,
+                )
+                return True
+
+            # We can enable stage-level scheduling
+            return False
+
+        # CPU training doesn't require stage-level scheduling
+        return True
+
+    def _try_stage_level_scheduling(self, rdd: RDD) -> RDD:
+        """Try to enable stage-level scheduling"""
+
+        if self._skip_stage_level_scheduling():
+            return rdd
+
+        ss = _get_spark_session()
+
+        # executor_cores will not be None
+        executor_cores = ss.sparkContext.getConf().get("spark.executor.cores")
+        assert executor_cores is not None
+
+        # Each training task requires cpu cores > total executor cores//2 + 1 which can
+        # make the tasks be sent to different executors.
+        #
+        # Please note that we can't set task_cpu_cores to the value which is smaller than
+        # total executor cores/2 because only task_gpu_amount can't make sure the tasks be
+        # sent to different executor even task_gpus=1.0
+
+        # Spark-rapids is a project to leverage GPUs to accelerate spark SQL and it can
+        # really help the performance. If spark-rapids is enabled. we don't allow other
+        # ETL gpu tasks running alongside training tasks to avoid OOM
```

Comment:
I'm a bit confused by this comment. Let me try to rephrase it and see if I got it right.

```
To send the tasks to different executors, the task must require at least `total executor cores // 2 + 1` CPU cores.

Without configuring the `task_cpu_cores` to a larger value, we can't make sure tasks are sent to different executors even if `task_gpu_amount` is set to 1.

When the `spark-rapids` plugin is enabled for SQL acceleration, we need to prevent other GPU-based ETL tasks from running during training to avoid OOM errors.
```

Reply:
Yes, you're correct.

Reply:
Could you please update the comments with a cleaner explanation?